File: Dopamine.Controls/TransitioningContentControl.vb
Imports System.Windows.Media.Animation
Imports System.Timers

Public Class TransitioningContentControl
    Inherits ContentControl

#Region "Variables"
    Private mTimer As Timer
#End Region

#Region "Dependency Properties"
    Public Shared ReadOnly FadeInProperty As DependencyProperty = DependencyProperty.Register("FadeIn", GetType(Boolean), GetType(TransitioningContentControl), New PropertyMetadata(Nothing))
    Public Shared ReadOnly FadeInTimeoutProperty As DependencyProperty = DependencyProperty.Register("FadeInTimeout", GetType(Double), GetType(TransitioningContentControl), New PropertyMetadata(Nothing))
    Public Shared ReadOnly SlideInProperty As DependencyProperty = DependencyProperty.Register("SlideIn", GetType(Boolean), GetType(TransitioningContentControl), New PropertyMetadata(Nothing))
    Public Shared ReadOnly SlideInTimeoutProperty As DependencyProperty = DependencyProperty.Register("SlideInTimeout", GetType(Double), GetType(TransitioningContentControl), New PropertyMetadata(Nothing))
    Public Shared ReadOnly SlideInFromProperty As DependencyProperty = DependencyProperty.Register("SlideInFrom", GetType(Integer), GetType(TransitioningContentControl), New PropertyMetadata(Nothing))
    Public Shared ReadOnly SlideInToProperty As DependencyProperty = DependencyProperty.Register("SlideInTo", GetType(Integer), GetType(TransitioningContentControl), New PropertyMetadata(Nothing))
    Public Shared ReadOnly RightToLeftProperty As DependencyProperty = DependencyProperty.Register("RightToLeft", GetType(Boolean), GetType(TransitioningContentControl), New PropertyMetadata(Nothing))
#End Region

#Region "Routed Events"
    Public Shared ReadOnly ContentChangedEvent As RoutedEvent = EventManager.RegisterRoutedEvent("ContentChanged", RoutingStrategy.Bubble, GetType(RoutedEventHandler), GetType(TransitioningContentControl))

    Public Custom Event ContentChanged As RoutedEventHandler
        AddHandler(ByVal value As RoutedEventHandler)
            Me.AddHandler(ContentChangedEvent, value)
        End AddHandler

        RemoveHandler(ByVal value As RoutedEventHandler)
            Me.RemoveHandler(ContentChangedEvent, value)
        End RemoveHandler

        RaiseEvent(ByVal sender As Object, ByVal e As RoutedEventArgs)
            Me.RaiseEvent(e)
        End RaiseEvent
    End Event

    Private Sub RaiseContentChangedEvent()
        Dim newEventArgs As New RoutedEventArgs(TransitioningContentControl.ContentChangedEvent)
        MyBase.RaiseEvent(newEventArgs)
    End Sub
#End Region

#Region "Properties"
    Public Property FadeIn As Boolean
        Get
            Return CBool(GetValue(FadeInProperty))
        End Get

        Set(ByVal value As Boolean)
            SetValue(FadeInProperty, value)
        End Set
    End Property

    Public Property FadeInTimeout As Double
        Get
            Return CDbl(GetValue(FadeInTimeoutProperty))
        End Get

        Set(ByVal value As Double)
            SetValue(FadeInTimeoutProperty, value)
        End Set
    End Property

    Public Property SlideIn As Boolean
        Get
            Return CBool(GetValue(SlideInProperty))
        End Get

        Set(ByVal value As Boolean)
            SetValue(SlideInProperty, value)
        End Set
    End Property

    Public Property SlideInTimeout As Double
        Get
            Return CDbl(GetValue(SlideInTimeoutProperty))
        End Get

        Set(ByVal value As Double)
            SetValue(SlideInTimeoutProperty, value)
        End Set
    End Property

    Public Property SlideInFrom As Integer
        Get
            Return CInt(GetValue(SlideInFromProperty))
        End Get

        Set(ByVal value As Integer)
            SetValue(SlideInFromProperty, value)
        End Set
    End Property

    Public Property SlideInTo As Integer
        Get
            Return CInt(GetValue(SlideInToProperty))
        End Get

        Set(ByVal value As Integer)
            SetValue(SlideInToProperty, value)
        End Set
    End Property

    Public Property RightToLeft As Boolean
        Get
            Return CBool(GetValue(RightToLeftProperty))
        End Get

        Set(ByVal value As Boolean)
            SetValue(RightToLeftProperty, value)
        End Set
    End Property
#End Region

#Region "Functions"
    Protected Overrides Sub OnContentChanged(oldContent As Object, newContent As Object)
        Me.DoAnimation()
    End Sub

    Private Sub DoAnimation()
        If Me.FadeInTimeout <> Nothing AndAlso Me.FadeIn Then
            Dim da As New DoubleAnimation
            da.From = 0
            da.To = 1
            da.Duration = New Duration(TimeSpan.FromSeconds(Me.FadeInTimeout))
            Me.BeginAnimation(OpacityProperty, da)
        End If

        If Me.SlideInTimeout <> Nothing AndAlso Me.SlideInTimeout > 0 AndAlso Me.SlideIn Then

            If Not Me.RightToLeft Then
                Dim ta As New ThicknessAnimation
                ta.From = New Thickness(Me.SlideInFrom, Me.Margin.Top, 2 * Me.SlideInTo - Me.SlideInFrom, Me.Margin.Bottom)
                ta.To = New Thickness(Me.SlideInTo, Me.Margin.Top, Me.SlideInTo, Me.Margin.Bottom)
                ta.Duration = New Duration(TimeSpan.FromSeconds(Me.SlideInTimeout))
                Me.BeginAnimation(MarginProperty, ta)
            Else
                Dim ta As New ThicknessAnimation
                ta.From = New Thickness(2 * Me.SlideInTo - Me.SlideInFrom, Me.Margin.Top, Me.SlideInFrom, Me.Margin.Bottom)
                ta.To = New Thickness(Me.SlideInTo, Me.Margin.Top, Me.SlideInTo, Me.Margin.Bottom)
                ta.Duration = New Duration(TimeSpan.FromSeconds(Me.SlideInTimeout))
                Me.BeginAnimation(MarginProperty, ta)
            End If
        End If

        If Me.mTimer IsNot Nothing Then
            Me.mTimer.Stop()
            RemoveHandler Me.mTimer.Elapsed, New ElapsedEventHandler(AddressOf Me.TimerElapsedHandler)
        End If

        Me.mTimer = New Timer

        Dim biggestTimeout As Double = Me.SlideInTimeout

        If Me.FadeInTimeout > Me.SlideInTimeout Then
            biggestTimeout = Me.FadeInTimeout
        End If

        Me.mTimer.Interval = TimeSpan.FromSeconds(biggestTimeout).TotalMilliseconds

        AddHandler Me.mTimer.Elapsed, New ElapsedEventHandler(AddressOf Me.TimerElapsedHandler)

        Me.mTimer.Start()
    End Sub

    Private Sub TimerElapsedHandler(sender As Object, e As ElapsedEventArgs)
        Me.mTimer.Stop()

        Try
            Application.Current.Dispatcher.BeginInvoke(Sub() Me.RaiseContentChangedEvent())
        Catch ex As Exception

        End Try
        
    End Sub
#End Region
End Class

File: Dopamine.Presentation/Controls/TransitioningContentControl.cs
using System;
using System.Timers;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media.Animation;

namespace Dopamine.Presentation.Controls
{
    public class TransitioningContentControl : ContentControl
    {
        private Timer timer;

        public static readonly DependencyProperty FadeInProperty = DependencyProperty.Register("FadeIn", typeof(bool), typeof(TransitioningContentControl), new PropertyMetadata(null));
        public static readonly DependencyProperty FadeInTimeoutProperty = DependencyProperty.Register("FadeInTimeout", typeof(double), typeof(TransitioningContentControl), new PropertyMetadata(null));
        public static readonly DependencyProperty SlideInProperty = DependencyProperty.Register("SlideIn", typeof(bool), typeof(TransitioningContentControl), new PropertyMetadata(null));
        public static readonly DependencyProperty SlideInTimeoutProperty = DependencyProperty.Register("SlideInTimeout", typeof(double), typeof(TransitioningContentControl), new PropertyMetadata(null));
        public static readonly DependencyProperty SlideInFromProperty = DependencyProperty.Register("SlideInFrom", typeof(int), typeof(TransitioningContentControl), new PropertyMetadata(null));
        public static readonly DependencyProperty SlideInToProperty = DependencyProperty.Register("SlideInTo", typeof(int), typeof(TransitioningContentControl), new PropertyMetadata(null));
        public static readonly DependencyProperty RightToLeftProperty = DependencyProperty.Register("RightToLeft", typeof(bool), typeof(TransitioningContentControl), new PropertyMetadata(null));

        public static readonly RoutedEvent ContentChangedEvent = EventManager.RegisterRoutedEvent("ContentChanged", RoutingStrategy.Bubble, typeof(RoutedEventHandler), typeof(TransitioningContentControl));
        
        public event RoutedEventHandler ContentChanged
        {
            add { this.AddHandler(ContentChangedEvent, value); }

            remove { this.RemoveHandler(ContentChangedEvent, value); }


        }
        private void RaiseContentChangedEvent()
        {
            RoutedEventArgs newEventArgs = new RoutedEventArgs(TransitioningContentControl.ContentChangedEvent);
            base.RaiseEvent(newEventArgs);
        }

        public bool FadeIn
        {
            get { return Convert.ToBoolean(GetValue(FadeInProperty)); }

            set { SetValue(FadeInProperty, value); }
        }

        public double FadeInTimeout
        {
            get { return Convert.ToDouble(GetValue(FadeInTimeoutProperty)); }

            set { SetValue(FadeInTimeoutProperty, value); }
        }

        public bool SlideIn
        {
            get { return Convert.ToBoolean(GetValue(SlideInProperty)); }

            set { SetValue(SlideInProperty, value); }
        }

        public double SlideInTimeout
        {
            get { return Convert.ToDouble(GetValue(SlideInTimeoutProperty)); }

            set { SetValue(SlideInTimeoutProperty, value); }
        }

        public int SlideInFrom
        {
            get { return Convert.ToInt32(GetValue(SlideInFromProperty)); }

            set { SetValue(SlideInFromProperty, value); }
        }

        public int SlideInTo
        {
            get { return Convert.ToInt32(GetValue(SlideInToProperty)); }

            set { SetValue(SlideInToProperty, value); }
        }

        public bool RightToLeft
        {
            get { return Convert.ToBoolean(GetValue(RightToLeftProperty)); }

            set { SetValue(RightToLeftProperty, value); }
        }

        protected override void OnContentChanged(object oldContent, object newContent)
        {
            this.DoAnimation();
        }

        private void DoAnimation()
        {
            if (this.FadeInTimeout != null && this.FadeIn)
            {
                var da = new DoubleAnimation();
                da.From = 0;
                da.To = 1;
                da.Duration = new Duration(TimeSpan.FromSeconds(this.FadeInTimeout));
                this.BeginAnimation(OpacityProperty, da);
            }


            if (this.SlideInTimeout != null && this.SlideInTimeout > 0 && this.SlideIn)
            {
                if (!this.RightToLeft)
                {
                    var ta = new ThicknessAnimation();
                    ta.From = new Thickness(this.SlideInFrom, this.Margin.Top, 2 * this.SlideInTo - this.SlideInFrom, this.Margin.Bottom);
                    ta.To = new Thickness(this.SlideInTo, this.Margin.Top, this.SlideInTo, this.Margin.Bottom);
                    ta.Duration = new Duration(TimeSpan.FromSeconds(this.SlideInTimeout));
                    this.BeginAnimation(MarginProperty, ta);
                }
                else
                {
                    var ta = new ThicknessAnimation();
                    ta.From = new Thickness(2 * this.SlideInTo - this.SlideInFrom, this.Margin.Top, this.SlideInFrom, this.Margin.Bottom);
                    ta.To = new Thickness(this.SlideInTo, this.Margin.Top, this.SlideInTo, this.Margin.Bottom);
                    ta.Duration = new Duration(TimeSpan.FromSeconds(this.SlideInTimeout));
                    this.BeginAnimation(MarginProperty, ta);
                }
            }

            if (this.timer != null)
            {
                this.timer.Stop();
                this.timer.Elapsed -= new ElapsedEventHandler(this.TimerElapsedHandler);
            }

            this.timer = new Timer();

            double biggestTimeout = this.SlideInTimeout;

            if (this.FadeInTimeout > this.SlideInTimeout)
            {
                biggestTimeout = this.FadeInTimeout;
            }

            this.timer.Interval = TimeSpan.FromSeconds(biggestTimeout).TotalMilliseconds;

            this.timer.Elapsed += new ElapsedEventHandler(this.TimerElapsedHandler);

            this.timer.Start();
        }

        private void TimerElapsedHandler(object sender, ElapsedEventArgs e)
        {
            this.timer.Stop();

            try
            {
                Application.Current.Dispatcher.BeginInvoke(new Action(() => this.RaiseContentChangedEvent()));
            }
            catch (Exception)
            {
            }
        }
    }
}
